src/tigerbeetle/client.zig

```diff
@@ -0,0 +1,839 @@
+const std = @import("std");
+const builtin = @import("builtin");
+const assert = std.debug.assert;
+
+const constants = @import("../constants.zig");
+const stdx = @import("../stdx.zig");
+
+const build_options = @import("vsr_options");
+
+const vsr = @import("vsr");
+const IO = vsr.io.IO;
+const MessagePool = vsr.message_pool.MessagePool;
+
+const tb = vsr.tigerbeetle;
+
+const MAX_SINGLE_REPL_INPUT = 10 * 4 * 1024;
+
+pub fn ClientType(comptime StateMachine: type, comptime MessageBus: type) type {
+    return struct {
+        const Self = @This();
+        const Client = vsr.Client(StateMachine, MessageBus);
+
+        fn fail(comptime fmt: []const u8, args: anytype) noreturn {
+            const stderr = std.io.getStdErr().writer();
+            stderr.print(fmt, args) catch unreachable;
+            std.os.exit(1);
+        }
+
+        fn print(comptime fmt: []const u8, args: anytype) void {
+            const stdout = std.io.getStdOut().writer();
+            stdout.print(fmt, args) catch return;
+        }
+
+        pub const Context = struct {
+            event_loop_done: bool,
+            request_done: bool,
+
+            repl: bool,
+            debug_logs: bool,
+
+            client: *Client,
+            message: ?*MessagePool.Message,
+
+            fn fail(context: *Context, comptime fmt: []const u8, args: anytype) void {
+                if (!context.repl) {
+                    Self.fail(fmt, args);
+                }
+
+                print(fmt, args);
+            }
+
+            fn debug(context: *Context, comptime fmt: []const u8, args: anytype) void {
+                if (context.debug_logs) {
+                    print("[Debug] " ++ fmt, args);
+                }
+            }
+
+            fn fail_at(
+                context: *Context,
+                input: []const u8,
+                index: usize,
+                comptime fmt: []const u8,
+                args: anytype,
+            ) void {
+                var line_number: usize = 1;
+                var column_number: usize = 0;
+
+                var line_beginning: usize = 0;
+                var found_current_line = false;
+
+                var i: usize = 0;
+                while (i < input.len) : (i += 1) {
+                    var c = input[i];
+
+                    if (c == '\n') {
+                        if (!found_current_line) {
+                            line_beginning = i + 1;
+                            line_number += 1;
+                            column_number = 0;
+                        } else {
+                            break;
+                        }
+                    } else {
+                        if (!found_current_line) {
+                            column_number += 1;
+                        }
+                    }
+
+                    if (i == index) {
+                        found_current_line = true;
+                    }
+                }
```

Comment:
Man, that is much simpler! I've written this so many times and never thought of that.

Reply:
Ah ok the one other thing I noticed this was doing before is that it also denotes the entire line where the error happens. So it can print that entire line and point at the column in the line where it occurs.

I think I need to revert for that.

Reply:
Hmm, in between the loops `cursor` is the start of the line. You could add a separate loop to find the end, which would give you the line in question. Though idk whether or not the end result would be better :shrug: